extensions/smallrye-reactive-messaging/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/deployment/WiringHelper.java

```diff
@@ -264,7 +264,10 @@ static boolean isSynthetic(MethodInfo method) {
     static Optional<AnnotationInstance> getAnnotation(TransformedAnnotationsBuildItem transformedAnnotations,
             InjectionPointInfo injectionPoint,
             DotName annotationName) {
-        Collection<AnnotationInstance> annotations = transformedAnnotations.getAnnotations(injectionPoint.getTarget());
+        // Query annotations for either field, or whole method
+        AnnotationTarget annotationTarget = injectionPoint.isField() ? injectionPoint.getAnnotationTarget()
```

Comment:
This is weird. We simplify the API but complicate its usage? Maybe the code below could be rewritten in order to leverage the new API?

Reply:
I don't follow, this is using `TransformedAnnotationsBuildItem` in order to gain access to already transformed annotations. However, Jandex in compatible mode (which it uses always) doesn't allow you to query for a `METHOD_PARAMETER` and you need to supply a method annotation target instead.
Best I can think of is improving `TransformedAnnotationsBuildItem` so that this happens automatically; is that what you had in mind?

Reply:
My point is that since we simplify we the API it should be less code, not more.

In this particular case you replaced:
`transformedAnnotations.getAnnotations(injectionPoint.getTarget())`
with 
`transformedAnnotations.getAnnotations(injectionPoint.isField() ? injectionPoint.getAnnotationTarget() : injectionPoint.getAnnotationTarget().asMethodParameter().method())`

Which is not good. So I wonder if the loop below could benefit from the new API?

Reply:
@mkouba I've added [third commit](https://github.com/quarkusio/quarkus/pull/40841/commits/2a3a41f312486f875095454b931eca45cb701ddb) which adjusts `TransformedAnnotationsBuildItem` to account for `AnnotationTarget` not being what Jandex expects. This also allows to simplify the code in `WiringHelper` back to what it was (minus the deprecated method invocation OFC).

Reply:
This `TransformedAnnotationsBuildItem` improvement is ok but if you look at the loop below you'll find out that this improvement is not needed at all and the logic that checks the position of the parameter is also useless because all you need to do is to iterate over annotations of a specific injection point and return the first one that matches the given name.

```java
static Optional<AnnotationInstance> getAnnotation(TransformedAnnotationsBuildItem transformedAnnotations,
            InjectionPointInfo injectionPoint,
            DotName annotationName) {
   // For field IP -> set of field annotations
   // For method param IP -> set of param annotations
   Collection<AnnotationInstance> annotations = transformedAnnotations.getAnnotations(injectionPoint.getAnnotationTarget());
   for (AnnotationInstance annotation : annotations) {
      if (annotationName.equals(annotation.name())) {
         return Optional.of(annotation);
      }
   }
   return Optional.empty();
}
```

Does that make sense?

Reply:
We discussed this on Zulip and the above code is incorrect as the underlying Jandex `AnnotationOverlay` won't work for method param. annotation target.
Possible solution for this case is to improve `TransformedAnnotationsBuildItem`  to add new methods (and deprecate current ones) which would automatically filter resulting annotations for method param annotation targets. I can upgrade the PR tomorrow with that.

Reply:
> Possible solution for this case is to improve TransformedAnnotationsBuildItem to add new methods (and deprecate current ones) which would automatically filter resulting annotations for method param annotation targets. I can upgrade the PR tomorrow with that.

:+1: